src/hotspot/share/opto/mulnode.cpp

```diff
@@ -239,18 +239,16 @@ MulNode* MulNode::make(Node* in1, Node* in2, BasicType bt) {
 //------------------------------Ideal------------------------------------------
 // Check for power-of-2 multiply, then try the regular MulNode::Ideal
 Node *MulINode::Ideal(PhaseGVN *phase, bool can_reshape) {
-  // Swap constant to right
   jint con;
-  if ((con = in(1)->find_int_con(0)) != 0) {
-    swap_edges(1, 2);
-    // Finish rest of method to use info in 'con'
-  } else if ((con = in(2)->find_int_con(0)) == 0) {
+  if ((con = in(2)->find_int_con(0)) == 0) {
+    // Call MulNode::Ideal() for common transformations. After that,
+    // the constant input should be on the right side.
     return MulNode::Ideal(phase, can_reshape);
   }
 
-  // Now we have a constant Node on the right and the constant in con
-  if (con == 0) return NULL;   // By zero is handled by Value call
```

Comment:
Value() runs after Ideal so why wouldn't we have con == 0 here?

Reply:
This Ideal function is returned in `if ((con = in(2)->find_int_con(0)) == 0)` block above if we have `con == 0`. So this line is actually unreachable.

Reply:
Ok. Can you reword the comment to state that explicitly?

Reply:
Done in latest commit

Reply:
Since there is now only a single assignment, I suggest to move the assignment out of the condition to make it more readable and more clear that we cannot have `con == 0` afterwards.
```
const jint con = in(2)->find_int_con(0);
if (con == 0) {
...
```
Same further down for `MulL`.